helix-view/src/document.rs

```diff
@@ -457,6 +459,101 @@ where
+            DocumentConfig {
+                config: Some(doc_config.config),
+                encoding: doc_config.encoding,
+                line_ending: doc_config.line_ending,
+            }
+        } else {
+            DocumentConfig::default()
+        }
+    }
+}
+
+/// Trait representing methods of configuring the document as it's being opened.
+trait ConfigureDocument {
+    type Config;
+    /// Loads document configuration for a file at `path`.
+    fn load(&self, path: &Path) -> Result<DocumentConfig<Self::Config>, Error>;
+    /// Applies document configuration except for the non-`config` fields of `DocumentConfig`.
+    fn configure_document(doc: &mut Document, settings: Self::Config);
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+struct Autodetect;
+#[derive(Clone, Copy, Debug, Default)]
+struct EditorConfig;
+
+impl ConfigureDocument for Autodetect {
+    type Config = ();
+
+    fn load(&self, _: &Path) -> Result<DocumentConfig<()>, Error> {
+        Ok(DocumentConfig::default())
+    }
+
+    fn configure_document(doc: &mut Document, _: Self::Config) {
+        doc.detect_indent();
+    }
+}
+
+impl ConfigureDocument for EditorConfig {
+    type Config = ec4rs::Properties;
+
+    fn load(&self, path: &Path) -> Result<DocumentConfig<Self::Config>, Error> {
+        use ec4rs::property::{Charset, EndOfLine};
+        use encoding::Encoding;
+        let mut config = ec4rs::properties_of(path)?;
+        config.use_fallbacks();
+        let encoding = config
+            .get_raw::<Charset>()
+            .filter_unset()
+            .into_result()
+            .ok()
+            .and_then(|string| Encoding::for_label(string.to_lowercase().as_bytes()));
+        let line_ending = match config.get::<EndOfLine>() {
+            Ok(EndOfLine::Lf) => Some(LineEnding::LF),
+            Ok(EndOfLine::CrLf) => Some(LineEnding::Crlf),
+            #[cfg(feature = "unicode-lines")]
+            Ok(EndOfLine::Cr) => Some(LineEnding::CR),
+            _ => None,
+        };
+        Ok(DocumentConfig {
+            config,
+            encoding,
+            line_ending,
+        })
+    }
+
+    fn configure_document(doc: &mut Document, settings: Self::Config) {
+        use ec4rs::property::{IndentSize, IndentStyle as IndentStyleEc, TabWidth};
+        match settings.get::<IndentStyleEc>() {
+            Ok(IndentStyleEc::Tabs) => doc.indent_style = IndentStyle::Tabs,
+            Ok(IndentStyleEc::Spaces) => {
+                let spaces = if let Ok(IndentSize::Value(cols)) = settings.get::<IndentSize>() {
+                    cols
+                } else {
+                    doc.tab_width
+                };
+                doc.indent_style = IndentStyle::Spaces(spaces.try_into().unwrap_or(u8::MAX));
+            }
+            _ => doc.detect_indent(),
+        }
+        if let Ok(TabWidth::Value(width)) = settings.get::<TabWidth>() {
+            doc.tab_width = width;
+        }
+    }
+}
```

Comment:
Why not use enum and this all could probably be simplified a lot. Should I just directly make the change and let you see what I mean?